### DISCUSSION TITLE
Using localhost with Unicorn?

### DISCUSSION THREAD

**Discussion by stevenharman** [socketry/localhost#16] [Q&A]
👋 Hello friends!

Has anyone succeeded, or even looked into, getting `localhost` working with Unicorn? If not, does anyone have pointers for getting started with such a effort?

It turns out googling "Unicorn localhost tls" (or "ssl") return a lot of results, but none of them that I've found are taking about _this_ localhost. 😆

**Comment by ioquatix** [comment]
The localhost has an incredibly simple interface, as long as you can provide a TLS context to the server, it should be possible to configure it.

``` ruby
# The (self-signed) authority to use:
hostname = "localhost"
authority = Localhost::Authority.fetch(hostname)
authority.server_context # OpenSSL::SSL::SSLContext
```

You just need to feed that into Unicorn's bind configuration. Is that possible? I checked the source code but I don't think Unicorn supports TLS.

**Comment by stevenharman** [comment]
> I checked the source code but I don't think Unicorn supports TLS.

I think you're right, sadly. It's been years since I've had to use Unicorn but I recall it being very optimized to run behind a proxy - typically `nginx`, or something like Heroku's router which handles TLS termination.

I wonder if I could at least use `localhost` to generate and manage the self-signed cert, so I don't have to muck with `openssl` directly... 🤔

**Reply by ioquatix** [reply]
Yes, you absolutely can do that. The certs are stored in `~/.localhost` but it's intended to be used via the interface provided.
